experiment/autobumper/main.go

```diff
@@ -0,0 +1,323 @@
+		variant   string
+		component string
+	}
+	versions := map[string][]delta{}
+	for image, newTag := range images {
+		if !strings.HasPrefix(image, prefix) {
+			continue
+		}
+		if strings.HasSuffix(image, ":"+newTag) {
+			continue
+		}
+		oldDate, oldCommit, oldVariant := bumper.DeconstructTag(tagFromName(image))
+		newDate, newCommit, _ := bumper.DeconstructTag(newTag)
+		k := oldCommit + ":" + newCommit
+		d := delta{
+			oldCommit: oldCommit,
+			newCommit: newCommit,
+			oldDate:   oldDate,
+			newDate:   newDate,
+			variant:   oldVariant,
+			component: componentFromName(image),
+		}
+		versions[k] = append(versions[k], d)
+	}
+
+	switch {
+	case len(versions) == 0:
+		return fmt.Sprintf("No %s changes.", name)
+	case len(versions) == 1 && summarise:
+		for k, v := range versions {
+			s := strings.Split(k, ":")
+			return fmt.Sprintf("%s changes: %s/compare/%s...%s (%s → %s)", name, repo, s[0], s[1], formatTagDate(v[0].oldDate), formatTagDate(v[0].newDate))
+		}
+	default:
+		changes := make([]string, 0, len(versions))
+		for k, v := range versions {
+			s := strings.Split(k, ":")
+			names := make([]string, 0, len(v))
+			for _, d := range v {
+				names = append(names, d.component+d.variant)
+			}
+			sort.Strings(names)
+			changes = append(changes, fmt.Sprintf("%s/compare/%s...%s | %s&nbsp;&#x2192;&nbsp;%s | %s",
+				repo, s[0], s[1], formatTagDate(v[0].oldDate), formatTagDate(v[0].newDate), strings.Join(names, ", ")))
+		}
+		sort.Slice(changes, func(i, j int) bool { return strings.Split(changes[i], "|")[1] < strings.Split(changes[j], "|")[1] })
+		return fmt.Sprintf("Multiple distinct %s changes:\n\nCommits | Dates | Images\n--- | --- | ---\n%s\n", name, strings.Join(changes, "\n"))
+	}
+	panic("unreachable!")
+}
+
+func getOncaller() (string, error) {
+	req, err := http.Get(oncallAddress)
+	if err != nil {
+		return "", err
+	}
+	defer req.Body.Close()
+	if req.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("HTTP error %d (%q) fetching current oncaller", req.StatusCode, req.Status)
+	}
+	oncall := struct {
+		Oncall struct {
+			TestInfra string `json:"testinfra"`
+		} `json:"Oncall"`
+	}{}
+	if err := json.NewDecoder(req.Body).Decode(&oncall); err != nil {
+		return "", err
+	}
+	return oncall.Oncall.TestInfra, nil
+}
+
+func generatePRBody(images map[string]string) string {
+	prowSummary := generateSummary("Prow", prowRepo, prowPrefix, true, images)
+	testImagesSummary := generateSummary("test-image", testImageRepo, testImagePrefix, false, images)
+	oncaller, err := getOncaller()
+
+	var assignment string
+	if err == nil {
+		if oncaller != "" {
+			assignment = "/cc @" + oncaller
+		} else {
+			assignment = "Nobody is currently oncall, so falling back to Blunderbuss."
+		}
+	} else {
+		assignment = fmt.Sprintf("An error occurred while finding an assignee: `%s`.\nFalling back to Blunderbuss.", err)
+	}
+	body := prowSummary + "\n\n" + testImagesSummary + "\n\n" + assignment + "\n"
+	return body
+}
+
+func main() {
```

Comment:
Is there anything in autobumper we want to test? Fine to do in later PRs.

Reply:
Probably the formatting logic, later.